### PR TITLE
More fixes found in tidy and libc++

### PR DESCRIPTION
**Pull request by cyyever**
cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen

### CODE REVIEW

File: aten/src/ATen/ParallelNative.cpp [hunk 1]
```diff
@@ -275,7 +275,7 @@ bool in_parallel_region() {
 #endif // C10_MOBILE
 }
 
-void intraop_launch(std::function<void()> func) {
+void intraop_launch(const std::function<void()>& func) {
 #ifndef C10_MOBILE
   if (!in_parallel_region() && get_num_threads() > 1) {
     _get_intraop_pool().run(std::move(func));
```
Comment (Skylion007) [inline aten/src/ATen/ParallelNative.cpp:281]: These should be by value considering it gets forwarded by C10_MOBILE code path. Be careful, clang-tidy won't check all paths of IFDEF

File: torch/csrc/api/include/torch/optim/optimizer.h [hunk 2]
```diff
@@ -91,9 +91,7 @@ class TORCH_API OptimizerParamGroup {
   const std::vector<Tensor>& params() const;
 
  protected:
-  // NOLINTNEXTLINE(cppcoreguidelines-non-private-member-variables-in-classes)
```
Comment (Skylion007) [inline torch/csrc/api/include/torch/optim/optimizer.h:94]: huh, these NOLINT look appropriate, the header is just not on the current include path.

File: aten/src/ATen/TensorIndexing.h [hunk 3]
```diff
@@ -111,15 +111,12 @@ TORCH_API std::ostream& operator<<(std::ostream& stream, const Slice& slice);
 // `torch.tensor([1, 2])`) | `torch::tensor({1, 2})`
 struct TORCH_API TensorIndex final {
   // Case 1: `at::indexing::None`
```
Comment (Skylion007) [inline aten/src/ATen/TensorIndexing.h:113]: Are we sure these are unneeded too? I don't see default initializers for things like boolean_ on all codepaths?

Comment (cyyever) [reply aten/src/ATen/TensorIndexing.h:113]: The members are
private:
  int64_t integer_ = 0;
  bool boolean_ = false;
  Slice slice_;
  Tensor tensor_;
  TensorIndexType type_;

POD types have in class initialisers and others use default constructor. So all members are initialised .

File: torch/csrc/api/include/torch/optim/optimizer.h (hunk 2 above)
Comment (cyyever) [reply torch/csrc/api/include/torch/optim/optimizer.h:94]: The reason is that we have put '-cppcoreguidelines-non-private-member-variables-in-classes' in the project clang-tidy config file. When I changed to those files, I would like to remove them. If you think we should put them back, I will revert the NOLINTs

File: aten/src/ATen/ParallelNative.cpp (hunk 1 above)
Comment (cyyever) [reply aten/src/ATen/ParallelNative.cpp:281]: After some thought, I think we should not change those thread pool API and I have reverted the related commits

**Comment by cyyever** [comment]
@Skylion007 I have discarded the problematic changes.

File: aten/src/ATen/core/ivalue.h [hunk 4]
```diff
@@ -1179,7 +1179,7 @@ struct TORCH_API IValue final {
       case Tag::Device:
         return false;
       case Tag::Stream:
-        return false;
+        return true;
```
Comment (Skylion007) [inline aten/src/ATen/core/ivalue.h:1182]: huh? Surprised by this change... Can't believe this didn't cause issues before...

**Review by Skylion007 — APPROVED** [review]

**Pull request closed without merging**